src/main/java/io/vertx/ext/mail/MailVerticle.java

```diff
@@ -0,0 +1,600 @@
+
+/*
+ first implementation of a SMTP client
+ */
+// TODO: this is not a verticle, the verticle is MailServiceVerticle
+// this class should probably be renamed
+
+/**
+ * @author <a href="http://oss.lehmann.cx/">Alexander Lehmann</a>
+ *
+ */
+public class MailVerticle {
+
+  private Vertx vertx;
+  private Handler<AsyncResult<JsonObject>> finishedHandler;
+
+  public MailVerticle(Vertx vertx, Handler<AsyncResult<JsonObject>> finishedHandler) {
+    this.vertx = vertx;
+    this.finishedHandler = finishedHandler;
+  }
+
+  private void shutdown() {
+    commandResultHandler = null;
+    socketShutDown = true;
+    if (ns != null) {
+      ns.close();
+      ns = null;
+    }
+    if (client != null) {
+      client.close();
+      client = null;
+    }
+  }
+
+  /*
+   * write command without masking anything
+   */
+  private void write(String str, Handler<AsyncResult<String>> commandResultHandler) {
+    write(str, -1, commandResultHandler);
+  }
+
+  /*
+   * write command masking everything after position blank
+   */
+  private void write(String str, int blank, Handler<AsyncResult<String>> commandResultHandler) {
+    this.commandResultHandler = commandResultHandler;
+    if(socketClosed) {
+      throwAsyncResult("connection was closed by server");
+    }
+    if (log.isDebugEnabled()) {
+      String logStr;
+      if (blank >= 0) {
+        StringBuilder sb = new StringBuilder();
+        for (int i = blank; i < str.length(); i++) {
+          sb.append('*');
+        }
+        logStr = str.substring(0, blank + 1) + sb;
+      } else {
+        logStr = str;
+      }
+      // avoid logging large mail body
+      if (logStr.length() < 1000) {
+        log.debug("command: " + logStr);
+      } else {
+        log.debug("command: " + logStr.substring(0, 1000) + "...");
+      }
+    }
+    ns.write(str + "\r\n");
+  }
+
+  private static final Logger log = LoggerFactory.getLogger(MailVerticle.class);
+  NetSocket ns;
+  boolean socketClosed;
+  boolean socketShutDown;
+
+  Handler<AsyncResult<String>> commandResultHandler;
+
+  private boolean capaStartTLS = false;
+  private Set<String> capaAuth = Collections.emptySet();
+  // 8BITMIME can be used if the server supports it, currently this is not
+  // implemented
+//  private boolean capa8BitMime = false;
+  // PIPELINING is not yet used
+//  private boolean capaPipelining = false;
+  private int capaSize = 0;
+
+  Email email;
+  String mailMessage;
+  String username;
+  String pw;
+  LoginOption login;
+  NetClient client;
+
+  public void sendMail(Email email, String username, String password, LoginOption login) {
+    this.email = email;
+    this.username = username;
+    pw = password;
+    this.login = login;
+
+    NetClientOptions netClientOptions = new NetClientOptions().setSsl(email.isSSLOnConnect());
+    client = vertx.createNetClient(netClientOptions);
+
+    client.connect(Integer.parseInt(email.getSmtpPort()), email.getHostName(), asyncResult -> {
```

Comment:
Yes, I didn't get to implementing this. It works like this and the overhead is small when using localhost, its a bit more overhead when logging in and when using e.g. googlemail, however this should not be a problem unless sending a lot of mails (e.g. when sending pw reset mails on a web page, it should be enough). I have that on the todo list (as a mental note at least, since I have not started a todo list yet)